velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window {
+
+namespace {
+
+// This function implementation is a generic way to compute any aggregation
+// used as a window function. It creates an Aggregate function object for the
+// window function invocation. At each row, we need to compute the aggregation
+// across all rows from the frameStart to frameEnd boundaries at that row.
+// This is done by invoking the singleGroup aggregation function methods
+// for each row.
+class AggregateWindowFunction : public exec::WindowFunction {
+ public:
+  AggregateWindowFunction(
+      const std::string& name,
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool) {
+    argTypes_.reserve(args.size());
+    argIndices_.reserve(args.size());
+    argVectors_.reserve(args.size());
+    for (const auto& arg : args) {
+      argTypes_.push_back(arg.type);
+      // TODO : Enhance code to handle constant arguments.
+      VELOX_CHECK_NULL(
+          arg.constantValue,
+          "Constant arguments are not supported in aggregate window functions.");
+      VELOX_CHECK(arg.index.has_value());
+      argIndices_.push_back(arg.index.value());
+      argVectors_.push_back(BaseVector::create(arg.type, 0, pool_));
+    }
+    // Create an Aggregate function object to do result computation. Window
+    // function usage only requires single group aggregation for calculating
+    // the function value for each row.
+    aggregate_ = exec::Aggregate::create(
+        name, core::AggregationNode::Step::kSingle, argTypes_, resultType);
+
+    // Aggregate initialization.
+    // Row layout is:
+    //  - null flags - one bit per aggregate.
+    //  - uint32_t row size,
+    //  - fixed-width accumulators - one per aggregate
+    //
+    // Here we always make space for a row size since we only have one
+    // row and no RowContainer. We also have a single aggregate here, so there
+    // is only one null bit.
+    singleGroupRowSize_ = kRowSizeOffset + sizeof(int32_t);
```

Comment:
kRowSizeOffset and kNullOffset are used only in this method; consider making them local static variables for readability.